File: backend/app/utils/files.py
"""Helpers for storing uploaded files (document scans, photos, logos)."""
from __future__ import annotations

import uuid
from pathlib import Path

from werkzeug.datastructures import FileStorage
from werkzeug.utils import secure_filename

from flask import current_app

ALLOWED_IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg", ".gif", ".webp", ".bmp", ".pdf"}


def _upload_dir() -> Path:
    path: Path = current_app.config["UPLOAD_DIR"]
    path.mkdir(parents=True, exist_ok=True)
    return path


def save_upload(file: FileStorage, subfolder: str = "") -> str:
    """Persist an uploaded file and return its path relative to UPLOAD_DIR.

    A random prefix is added to avoid collisions while keeping the original
    extension for previewing.
    """
    original = secure_filename(file.filename or "file")
    ext = Path(original).suffix.lower()
    if ext and ext not in ALLOWED_IMAGE_EXTENSIONS:
        from app.utils.errors import ValidationError

        raise ValidationError(f"Недопустимый тип файла: {ext}")

    target_dir = _upload_dir()
    if subfolder:
        target_dir = target_dir / secure_filename(subfolder)
        target_dir.mkdir(parents=True, exist_ok=True)

    filename = f"{uuid.uuid4().hex}{ext}"
    relative = filename if not subfolder else f"{subfolder}/{filename}"
    file.save(target_dir / filename)
    return relative


def absolute_path(relative: str | None) -> Path | None:
    """Resolve a stored relative path to an absolute filesystem path."""
    if not relative:
        return None
    return _upload_dir() / relative

File: backend-go/internal/media/files.go
// Package media handles file uploads, QR codes, PDF documents and Excel
// exports for the backend.
package media

import (
	"fmt"
	"io"
	"mime/multipart"
	"os"
	"path/filepath"
	"strings"

	"github.com/google/uuid"
)

// Generator produces media artefacts (uploads, QR, PDF, Excel) into the
// configured upload and export directories.
type Generator struct {
	UploadDir string
	ExportDir string
}

// NewGenerator builds a media generator bound to the given directories.
func NewGenerator(uploadDir, exportDir string) *Generator {
	return &Generator{UploadDir: uploadDir, ExportDir: exportDir}
}

var allowedUploadExt = map[string]bool{
	".png": true, ".jpg": true, ".jpeg": true, ".gif": true,
	".webp": true, ".bmp": true, ".pdf": true,
}

// SaveUpload persists an uploaded file and returns its path relative to the
// upload directory. A random prefix avoids collisions while keeping the
// original extension.
func (g *Generator) SaveUpload(fileHeader *multipart.FileHeader, subfolder string) (string, error) {
	ext := strings.ToLower(filepath.Ext(fileHeader.Filename))
	if ext != "" && !allowedUploadExt[ext] {
		return "", fmt.Errorf("недопустимый тип файла: %s", ext)
	}

	subfolder = sanitizeSubfolder(subfolder)
	targetDir := g.UploadDir
	if subfolder != "" {
		targetDir = filepath.Join(g.UploadDir, subfolder)
	}
	if err := os.MkdirAll(targetDir, 0o755); err != nil {
		return "", err
	}

	filename := uuid.NewString() + ext
	dst := filepath.Join(targetDir, filename)

	src, err := fileHeader.Open()
	if err != nil {
		return "", err
	}
	defer src.Close()

	out, err := os.Create(dst)
	if err != nil {
		return "", err
	}
	defer out.Close()

	if _, err := io.Copy(out, src); err != nil {
		return "", err
	}

	if subfolder == "" {
		return filename, nil
	}
	return subfolder + "/" + filename, nil
}

// AbsoluteUploadPath resolves a stored relative path to an absolute filesystem
// path inside the upload directory.
func (g *Generator) AbsoluteUploadPath(relative string) string {
	return filepath.Join(g.UploadDir, filepath.Clean("/"+relative))
}

// AbsoluteExportPath resolves a stored relative path inside the export dir.
func (g *Generator) AbsoluteExportPath(relative string) string {
	return filepath.Join(g.ExportDir, filepath.Clean("/"+relative))
}

func sanitizeSubfolder(sub string) string {
	sub = strings.TrimSpace(sub)
	sub = strings.ReplaceAll(sub, "..", "")
	sub = strings.Trim(sub, "/\\")
	return sub
}
